src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -241,6 +280,240 @@ void DiscoveryDataBase::AckedFunctor::operator () (
+
+    auto wit = writers_.find(writer_guid);
+    if (wit != writers_.end())
+    {
+        wit->second.set_disposal(ch);
+    }
+
+    auto pit = participants_.find(writer_guid.guidPrefix);
+    if (pit != participants_.end())
+    {
+        pit->second.remove_writer(writer_guid);
+    }
+
+    for (auto tit = writers_by_topic_.begin(); tit != writers_by_topic_.end(); ++tit)
+    {
+        tit->second.erase(
+                std::remove(tit->second.begin(), tit->second.end(), writer_guid),
+                tit->second.end());
+    }
+
+    if(std::find(disposals_.begin(), disposals_.end(), ch) == disposals_.end())
+    {
+        disposals_.push_back(ch);
+    }
+
+}
+
+void DiscoveryDataBase::process_dispose_reader(
+        eprosima::fastrtps::rtps::CacheChange_t* ch)
+{
+    const eprosima::fastrtps::rtps::GUID_t& reader_guid = guid_from_change(ch);
+
+    auto rit = readers_.find(reader_guid);
+    if (rit != readers_.end())
+    {
+        rit->second.set_disposal(ch);
+    }
+
+    auto pit = participants_.find(reader_guid.guidPrefix);
+    if (pit != participants_.end())
+    {
+        pit->second.remove_reader(reader_guid);
+    }
+
+    for (auto tit = readers_by_topic_.begin(); tit != readers_by_topic_.end(); ++tit)
+    {
+        tit->second.erase(
+                std::remove(tit->second.begin(), tit->second.end(), reader_guid),
+                tit->second.end());
+    }
+
+    if(std::find(disposals_.begin(), disposals_.end(), ch) == disposals_.end())
+    {
+        disposals_.push_back(ch);
+    }
+}
+
+void DiscoveryDataBase::process_dispose_participant(
+        eprosima::fastrtps::rtps::CacheChange_t* ch)
+{
+    const eprosima::fastrtps::rtps::GUID_t& participant_guid = guid_from_change(ch);
+
+    // Change DATA(p) with DATA(Up) in participants map
+    auto pit = participants_.find(participant_guid.guidPrefix);
+    if (pit != participants_.end())
+    {
+        pit->second.set_disposal(ch);
+    }
+
+    // Delete entries from writers_/readers_ belonging to the participant
+    for (auto wit = writers_.begin(); wit != writers_.end(); ++wit)
+    {
+        if (wit->first.guidPrefix == participant_guid.guidPrefix)
+        {
+            writers_.erase(wit->first);
+            --wit;
+        }
+    }
+
+    for (auto rit = readers_.begin(); rit != readers_.end(); ++rit)
+    {
+        if (rit->first.guidPrefix == participant_guid.guidPrefix)
+        {
+            readers_.erase(rit->first);
+            --rit;
+        }
+    }
+
+    // Delete Participant entries from readers_by_topic and writers_by_topic
+    for (auto tit = readers_by_topic_.begin(); tit != readers_by_topic_.end(); ++tit)
+    {
+        for (auto rit = tit->second.begin(); rit != tit->second.end(); ++rit)
+        {
+            if(rit->guidPrefix == participant_guid.guidPrefix)
+            {
+                tit->second.erase(rit);
+                --rit;
+            }
```

Comment:
Maybe we could check here whether the list of readers in that topic is empty, and if it is delete the entry of that topic from `readers_by_topic_`. Same goes for `writer_by_topic_`